
####################
# GBS masters data #
####################
###############################################################
# Underwood Col/Ler F2 2016 lane1 ind=96 cos=729 cos.ind=7.59 WTlane1.*.hqmask.smooth.co.txt #
# Underwood Col/Ler F2 2016 lane2 ind=96 cos=751 cos.ind=7.82 WTlane2.*.hqmask.smooth.co.txt #
# Serra Col/Ler F2 2016 lane1 ind=82 cos=611 cos.ind=7.45 n96.HS1.wtF21.bchqsnvmask.smooth.co.txt #####
# Serra Col/Ler F2 2016 lane2 ind=89 cos=658 cos.ind=7.39 n96.HS2.wtF21.bchqsnvmask.smooth.co.txt #
# Serra Col/Ler F2 2016 lane3 ind=74 cos=571 cos.ind=7.72 n96.HS.WT3F21.bchqsnvmask.smooth.co.txt ####
# Serra SA Col/Ler F2 2016 lane1 ind=96 cos=714 cos.ind=7.43 n96.HS1.SAF21.bchqsnvmask.smooth.co.txt #
# Serra SA Col/Ler F2 2016 lane2 ind=96 cos=723 cos.ind=7.53 n96.HS.SA21.bchqsnvmask.smooth.co.txt #
# Serra JA Col/Ler F2 2016 lane1 ind=85 cos=601 cos.ind=7.07 n96.HS1.JAF21.bchqsnvmask.smooth.co.txt # ind=84 (lib 58 cos=0)
# Serra JA Col/Ler F2 2016 lane2 ind=87 cos=653 cos.ind=7.51 n96.HS.JA2F21.bchqsnvmask.smooth.co.txt ##########
# Underwood HEI10 Col/Ler F2 2016 lane1 ind=96 cos=1456 cos.ind=15.17 n96.Hei10F21.bchqsnvmask.smooth.co.txt #
# Underwood HEI10 Col/Ler F2 2016 lane2 ind=96 cos=1472 cos.ind=15.33 n96.hei10lane2F21.bchqsnvmask.smooth.co.txt #
# Underwood cmt3 Col/Ler F2 2016 lane1 ind=96 cos=708 cos.ind=7.34 n96.Cmt3F21.bchqsnvmask.smooth.co.txt ####
# Underwood cmt3 Col/Ler F2 2016 lane2 ind=96 cos=702 cos.ind=7.31 n96.Cmt3.lane2.F21.bchqsnvmask.smooth.co.txt #
# Underwood cmt3 Col/Ler F2 2016 lane3 ind=96 cos=702 cos.ind=7.31 n96.cmt331.bchqsnvmask.smooth.co.txt #
# Underwood cmt3 Col/Ler F2 2016 lane4 ind=96 cos=691 cos.ind=7.20 n96.cmt3410.bchqsnvmask.smooth.co.txt ########
# Serra recq4a recq4b Col/Ler F2 2017 lane1 ind=96 cos=2396 cos.ind=24.96 n96.HS.Recq.series11.bchqsnvmask.smooth.co.txt #
# Serra recq4a recq4b Col/Ler F2 2017 lane2 ind=95 cos=2387 cos.ind=25.13  n96_HS_Recq_series21.bchqsnvmask.smooth.co.txt #######
# Serra recq4a recq4b HEI10 Col/Ler F2 2017 lane1 ind=96 cos=3060 cos.ind=31.88 n96.HS.Recq.HEI10.series11.bchqsnvmask.smooth.co.txt # cos=3036
# Serra recq4a recq4b HEI10 Col/Ler F2 2017 lane2 ind=96 cos=2888 cos.ind=30.08 n96.HS.Recq.HEI10.series210.bchqsnvmask.smooth.co.txt # cos=2872
# Lambing recq4a recq4b cmt3 Col/Ler F2 2017 lane1 ind=96 cos=2543 cos.ind=26.49 n96.CL.recq.cmt3.lane1.1.bchqsnvmask.smooth.co.txt #
# Lambing recq4a recq4b cmt3 Col/Ler F2 2017 lane2 ind=96 cos=2524 cos.ind=26.29 n96.CL.RC2F2.1.bchqsnvmask.smooth.co.txt #
# Lambing HEI10 cmt3 Col/Ler F2 2017 lane1 ind=96 cos=1476 cos.ind=15.38 n96.CL.HEI10.cmt3.lane1.1.bchqsnvmask.smooth.co.txt #########
# Lambing HEI10 cmt3 Col/Ler F2 2017 lane2 ind=96 cos=1409 cos.ind=14.68 n96.CL.HC2F2.2.bchqsnvmask.smooth.co.txt #
# Lambing asy1/+ Col/Ler F2 2018 lane1 ind=96 cos=768 cos.ind=8 n96.CL.asy1hete.series1.1.bchqsnvmask.smooth.co.txt #######
# Lambing asy1/+ Col/Ler F2 2018 lane2 ind=95 cos=744 cos.ind=7.83 n96.CL.asy1hete.series2.68.bchqsnvmask.smooth.co.txt # ind=96 cos=768
# Lambing Col/WS lane1 colws.lane1.1.mask.smooth.co.txt # ind=96 cos=753 
## Lambing Col/WS lane2 not done yet?
# Lawrence Col/Bur lane1 colburwt.lane1.1.mask.smooth.co.txt # ind=96 cos=750
## Lawrence Col/Bur lane2 not done yet
# Lawrence Col/Bur taf4b lane1 taf4b.lane1.1.mask.smooth.co.txt # ind=96 cos=647
# Blackwell Col/Cvi lane1 colcvi.lane1.2.mask.smooth.co.txt # ind=96 cos=952
## Blackwell Col/Cvi lane2 not done yet
# Blackwell Col/Cvi msh2 lane1 colcvi.msh2.lane1.1.mask.smooth.co.txt # ind=96 cos=1174
## Blackwell Col/Cvi msh2 lane2 not done yet
# Blackwell Col/Ler msh2 lane1 msh2.ler1.1.bchqsnvmask.smooth.co.txt # ind=96 cos=958
# Blackwell Col/Ler msh2 lane2 msh2.ler2.53.bchqsnvmask.smooth.co.txt # ind=95 cos=915
# Lambing Col/Ler sni1 lane1 n96.CL.sni1.series1.1.bchqsnvmask.smooth.co.txt # ind=94 cos=1030
# Yelina Col/Ler BC met1 lane1 n96.met11.bchqsnvmask.smooth.co.txt # ind=96 cos=1083
# Yelina Col/Ler BC met1 lane2 n96.met1.lane2.1.bchqsnvmask.smooth.co.txt # ind=96 cos=543
# Yelina Col/Ler BC lane1 n96.nywtbc.lane1.1.bchqsnvmask.smooth.co.txt # ind=92 cos=517
# Yelina Col/Ler BC lane2 n96.nywtbc.lane2.95.bchqsnvmask.smooth.co.txt # ind=96 cos=503
##########################################################################

###############################################################################
# asy1 library 2.82 has 24 crossovers - way off so unlikely to be real - drop #
###############################################################################

lib.nums <- seq(1,96,by=1)
asy1.cos <- NULL
for(k in 1:length(lib.nums)){
	print(k)
	dat <- read.table(file=paste("n96.CL.asy1hete.series1.",k,".bchqsnvmask.smooth.co.txt",sep=""),header=F)
	all <- NULL
	for(i in 1:5){	
		print(i)
		chr <- dat[which(dat[,2]==i),]
		if((length(chr[,1])>1)==T){
			start <- chr[,4]
			stop <- chr[,3]
			stop <- stop[-1]
			start <- start[-length(start)]
			diff <- stop-start		
			mid <- round(diff/2)
			cos <- start+mid
			chrs <- rep(i,length(cos))
			lib <- rep(paste("1.",lib.nums[k],sep=""),length(chrs))
			width <- stop-start		
			bind <- cbind(lib,chrs,start,stop,cos,width)	
			all <- rbind(all,bind)	
			print(dim(all))
			}
		}
	asy1.cos <- rbind(asy1.cos,all)
}
for(k in 1:length(lib.nums)){
	print(k)
	dat <- read.table(file=paste("n96.CL.asy1hete.series2.",k,".bchqsnvmask.smooth.co.txt",sep=""),header=F)
	all <- NULL
	for(i in 1:5){	
		print(i)
		chr <- dat[which(dat[,2]==i),]
		if((length(chr[,1])>1)==T){
			start <- chr[,4]
			stop <- chr[,3]
			stop <- stop[-1]
			start <- start[-length(start)]
			diff <- stop-start		
			mid <- round(diff/2)
			cos <- start+mid
			chrs <- rep(i,length(cos))
			lib <- rep(paste("2.",lib.nums[k],sep=""),length(chrs))
			width <- stop-start		
			bind <- cbind(lib,chrs,start,stop,cos,width)	
			all <- rbind(all,bind)	
			print(dim(all))
			}
		}
	asy1.cos <- rbind(asy1.cos,all)
}
print(dim(asy1.cos))
asy1.cos <- asy1.cos[-which(asy1.cos[,1]=="2.82"),]

mask.cos <- read.table(file="all.maskcos.txt")
hswt.cos <- mask.cos[which(mask.cos[,8]=="hs.wt"),]

wt.ind <- 245
asy1.ind <- 191

all.ind <- NULL
for(i in 1:3){
	print(i)
	lane <- hswt.cos[which(hswt.cos[,1]==i),]
	lane.libs <- unique(lane[,2])
	cos.ind <- NULL
	for(k in 1:length(lane.libs)){
		print(k)
		cos.ind <- c(cos.ind,length(which(lane[,2]==lane.libs[k])))
	}
	all.ind <- c(all.ind,cos.ind)
}
wt.ind <- all.ind

uni <- unique(asy1.cos[,1])
all.ind <- NULL
for(k in 1:length(uni)){
	print(i)
	sel <- asy1.cos[which(asy1.cos[,1]==uni[k]),]
	all.ind <- c(all.ind,length(sel[,1]))
}
asy1.ind <- all.ind

par(mfcol=c(1,2))
par(mar=c(1.8,1.8,1.8,1.8))
hist(wt.ind,breaks=100,xlim=c(0,15),main="wt")
abline(v=mean(wt.ind),col=2,lty=2,lwd=0.5)
hist(asy1.ind,breaks=100,xlim=c(0,15),main="asy1/+")
abline(v=mean(asy1.ind),col=2,lty=2,lwd=0.5)

t.test(wt.ind,asy1.ind)
95 percent confidence interval:
 -0.82570542  0.01365285
sample estimates:
mean of x mean of y 
 7.510204  7.916230 
 
dat <- wt.ind
error <- qt(0.975,df=length(dat)-1)*sd(dat)/sqrt(length(dat))
lower <- mean(dat)-error
upper <- mean(dat)+error
print(lower)
7.225
print(mean(dat))
7.510
print(upper)
7.795

dat <- asy1.ind
error <- qt(0.975,df=length(dat)-1)*sd(dat)/sqrt(length(dat))
lower <- mean(dat)-error
upper <- mean(dat)+error
print(lower)
7.607
print(mean(dat))
7.916
print(upper)
8.226

##########################################################
# parse GBS genotypes along chromosomes into rQTL format #
##########################################################

map.pwd <- "/Users/IanHenderson/Desktop/Lab_records/papers/Serra_PNAS/Fig_1_GBS/masters/"
chrs <- c("Chr1","Chr2","Chr3","Chr4","Chr5")
chr.ends <- c(30427671,19698289,23459830,18585056,26975502)
all.chr.coords <- NULL
all.chr.lab <- NULL
for(i in 1:5){
	print(i)
	chr.coords <- c(seq(1,chr.ends[i],by=1000000),chr.ends[i])
	all.chr.coords <- c(all.chr.coords,chr.coords)
	chr.lab <- rep(i,length(chr.coords))
	all.chr.lab <- c(all.chr.lab,chr.lab)
}
one.samp <- seq(437,532,by=1)
two.samp <- seq(533,628,by=1)

pop.asy11.states <- NULL
for(k in 1:96){
	print(k)		
	asy11.dat <- read.table(file=paste("n96.CL.asy1hete.series1.",k,".bchqsnvmask.smooth.co.txt",sep=""),header=F)
	print(k)
	all.asy11.states <- NULL
	for(i in 1:5){
		print(i)
		chr.coords <- c(seq(1,chr.ends[i],by=1000000),chr.ends[i])
		genoasy11.states <- rep(0,length(chr.coords))
		asy11.chr <- asy11.dat[which(asy11.dat[,2]==i),]
		for(m in 1:length(asy11.chr[,1])){
			print(m)
			genoasy11.states[which(chr.coords>=asy11.chr[m,3]&chr.coords<=asy11.chr[m,4])] <- asy11.chr[m,5]
		}		
		genoasy11.states[which(genoasy11.states==1)] <- "A"
		genoasy11.states[which(genoasy11.states==2)] <- "H"
		genoasy11.states[which(genoasy11.states==3)] <- "B"
		all.asy11.states <- c(all.asy11.states,genoasy11.states)
		}
	pop.asy11.states <- rbind(pop.asy11.states,all.asy11.states)
}
pop.asy12.states <- NULL
lib.nums <- c(seq(1,81,by=1),seq(83,96,by=1))
for(k in 1:length(lib.nums)){
	print(k)
	asy12.dat <- read.table(file=paste("n96.CL.asy1hete.series2.",lib.nums[k],".bchqsnvmask.smooth.co.txt",sep=""),header=F)
   	print(k)
	all.asy12.states <- NULL
	for(i in 1:5){
		print(i)
		chr.coords <- c(seq(1,chr.ends[i],by=1000000),chr.ends[i])
		genoasy12.states <- rep(0,length(chr.coords))
		asy12.chr <- asy12.dat[which(asy12.dat[,2]==i),]
		for(m in 1:length(asy12.chr[,1])){
			print(m)
			genoasy12.states[which(chr.coords>=asy12.chr[m,3]&chr.coords<=asy12.chr[m,4])] <- asy12.chr[m,5]
		}		
		genoasy12.states[which(genoasy12.states==1)] <- "A"
		genoasy12.states[which(genoasy12.states==2)] <- "H"
		genoasy12.states[which(genoasy12.states==3)] <- "B"
		all.asy12.states <- c(all.asy12.states,genoasy12.states)
	}
	pop.asy12.states <- rbind(pop.asy12.states,all.asy12.states)
}
pop.asy1.dat <- rbind(all.chr.coords,all.chr.lab,pop.asy11.states,pop.asy12.states)
header.col <- c("id","",seq(1,length(pop.asy1.dat[,1])-2,by=1))
pop.asy1.dat <- cbind(header.col,pop.asy1.dat)
pop.asy1.dat <- as.data.frame(pop.asy1.dat,row.names=F,col.names=F)
write.table(pop.asy1.dat,file="asy1.f2mask.csv",row.names=F,col.names=F,sep=",")

lib.nums <- c(seq(1,7,by=1),seq(9,10,by=1),seq(13,17,by=1),seq(19,25,by=1),seq(29,30,by=1),seq(32,38,by=1),seq(40,50,by=1),seq(52,56,by=1),seq(58,73,by=1),seq(75,86,by=1),seq(88,92,by=1),seq(94,96,by=1))
pop.hswt1.states <- NULL
for(k in 1:length(lib.nums)){
	print(k)
	hswt1.dat <- read.table(file=paste(map.pwd,"n96.HS1.wtF2",lib.nums[k],".bchqsnvmask.smooth.co.txt",sep=""),header=F)
   	print(k)
	all.hswt1.states <- NULL
	for(i in 1:5){
		print(i)
		chr.coords <- c(seq(1,chr.ends[i],by=1000000),chr.ends[i])
		genohswt1.states <- rep(0,length(chr.coords))
		hswt1.chr <- hswt1.dat[which(hswt1.dat[,2]==i),]
		for(m in 1:length(hswt1.chr[,1])){
			print(m)
			genohswt1.states[which(chr.coords>=hswt1.chr[m,3]&chr.coords<=hswt1.chr[m,4])] <- hswt1.chr[m,5]
		}		
		genohswt1.states[which(genohswt1.states==1)] <- "A"
		genohswt1.states[which(genohswt1.states==2)] <- "H"
		genohswt1.states[which(genohswt1.states==3)] <- "B"
		all.hswt1.states <- c(all.hswt1.states,genohswt1.states)
	}
	pop.hswt1.states <- rbind(pop.hswt1.states,all.hswt1.states)
}
lib.nums <- c(seq(1,2,by=1),seq(4,6,by=1),seq(8,42,by=1),seq(44,55,by=1),seq(57,70,by=1),seq(72,75,by=1),seq(77,91,by=1),seq(93,96,by=1))
pop.hswt2.states <- NULL
for(k in 1:length(lib.nums)){
	print(k)
	hswt2.dat <- read.table(file=paste(map.pwd,"n96.HS2.wtF2",lib.nums[k],".bchqsnvmask.smooth.co.txt",sep=""),header=F)
	all.hswt2.states <- NULL
	for(i in 1:5){
		print(i)
		chr.coords <- c(seq(1,chr.ends[i],by=1000000),chr.ends[i])
		genohswt2.states <- rep(0,length(chr.coords))
		hswt2.chr <- hswt2.dat[which(hswt2.dat[,2]==i),]
		for(m in 1:length(hswt2.chr[,1])){
			print(m)
			genohswt2.states[which(chr.coords>=hswt2.chr[m,3]&chr.coords<=hswt2.chr[m,4])] <- hswt2.chr[m,5]
		}		
		genohswt2.states[which(genohswt2.states==1)] <- "A"
		genohswt2.states[which(genohswt2.states==2)] <- "H"
		genohswt2.states[which(genohswt2.states==3)] <- "B"
		all.hswt2.states <- c(all.hswt2.states,genohswt2.states)
	}
	pop.hswt2.states <- rbind(pop.hswt2.states,all.hswt2.states)
}
lib.nums <- c(seq(1,61,by=1),seq(68,80,by=1))
pop.hswt3.states <- NULL
for(k in 1:length(lib.nums)){
	print(k)
	hswt3.dat <- read.table(file=paste(map.pwd,"n96.HS.WT3F2",lib.nums[k],".bchqsnvmask.smooth.co.txt",sep=""),header=F)
	all.hswt3.states <- NULL
	for(i in 1:5){
		print(i)
		chr.coords <- c(seq(1,chr.ends[i],by=1000000),chr.ends[i])
		genohswt3.states <- rep(0,length(chr.coords))
		hswt3.chr <- hswt3.dat[which(hswt3.dat[,2]==i),]
		for(m in 1:length(hswt3.chr[,1])){
			print(m)
			genohswt3.states[which(chr.coords>=hswt3.chr[m,3]&chr.coords<=hswt3.chr[m,4])] <- hswt3.chr[m,5]
		}		
		genohswt3.states[which(genohswt3.states==1)] <- "A"
		genohswt3.states[which(genohswt3.states==2)] <- "H"
		genohswt3.states[which(genohswt3.states==3)] <- "B"
		all.hswt3.states <- c(all.hswt3.states,genohswt3.states)
	}
	pop.hswt3.states <- rbind(pop.hswt3.states,all.hswt3.states)
}
pop.hswt.dat <- rbind(all.chr.coords,all.chr.lab,pop.hswt1.states,pop.hswt2.states,pop.hswt3.states)
header.col <- c("id","",seq(1,length(pop.hswt.dat[,1])-2,by=1))
pop.hswt.dat <- cbind(header.col,pop.hswt.dat)
pop.hswt.dat <- as.data.frame(pop.hswt.dat,row.names=F,col.names=F)
write.table(pop.hswt.dat,file="hswt.f2mask.csv",row.names=F,col.names=F,sep=",")

###################################
# analyse genetic maps using rQTL #
###################################

library(qtl)
hswt.map <- read.cross(file="hswt.f2mask.csv",genotypes=c("A","H","B"))
asy1.map <- read.cross(file="asy1.f2mask.csv",genotypes=c("A","H","B"))

par(mfcol=c(1,2))
par(mar=c(1.8,1.8,1.8,1.8))
plot.rf(hswt.map,main="wt hs rf")
plot.rf(asy1.map,main="asy1 rf")

wt.tot <- 245
asy1.tot <- 191

mean(wt.ind)
7.51
sd(wt.ind)
2.26

mean(asy1.ind)
7.92
sd(asy1.ind)
2.17

par(mfcol=c(1,2))
hist(wt.ind,breaks=1000,main="cos per wt F2")
hist(asy1.ind,breaks=1000,main="cos per asy1/+ F2")

hswt.haldane.map <- read.cross(file="hswt.f2mask.csv",genotypes=c("A","H","B"),map.function="haldane")
count.XO <- (sum(colSums(countXO(hswt.haldane.map,bychr=T))))/wt.tot
# 7.44
asy1.haldane.map <- read.cross(file="asy1.f2mask.csv",genotypes=c("A","H","B"),map.function="haldane")
count.XO <- (sum(colSums(countXO(asy1.haldane.map,bychr=T))))/asy1.tot
# 7.86

plotMap(hswt.haldane.map,asy1.haldane.map,main="wt vs asy1/+ haldane",ylim=c(115,0))

wt.geno <- hswt.map$geno
wtchr.len <- NULL
for(i in 1:5){
	chr <- wt.geno[[i]]
	map <- chr$map
	wtchr.len <- c(wtchr.len,map[length(map)])
}
30427671 19698289 23459830 18585056 26975502 
93.73843 71.63953 74.72704 66.97380 90.52054 
wttot.len <- round(sum(wtchr.len))
398

asy1.geno <- asy1.map$geno
asy1chr.len <- NULL
for(i in 1:5){
	chr <- asy1.geno[[i]]
	map <- chr$map
	asy1chr.len <- c(asy1chr.len,map[length(map)])
}
30427671 19698289 23459830 18585056 26975502 
98.24693 76.51397 78.78950 70.89854 97.43891 
asy1tot.len <- round(sum(asy1chr.len))
422

library(vcd)
wt.gf <- goodfit(wt.ind,type="poisson",method="ML")
asy1.gf <- goodfit(asy1.ind,type="poisson",method="ML")
plot(wt.gf,main="goodfit_wt",xlim=c(0,18))
plot(asy1.gf,main="goodfit_asy1/+",xlim=c(0,18))

##################################
# plotting along the chromosomes #
##################################

lib.nums <- seq(1,96,by=1)
asy1.cos <- NULL
for(k in 1:length(lib.nums)){
	print(k)
	dat <- read.table(file=paste("n96.CL.asy1hete.series1.",k,".bchqsnvmask.smooth.co.txt",sep=""),header=F)
	all <- NULL
	for(i in 1:5){	
		print(i)
		chr <- dat[which(dat[,2]==i),]
		if((length(chr[,1])>1)==T){
			start <- chr[,4]
			stop <- chr[,3]
			stop <- stop[-1]
			start <- start[-length(start)]
			diff <- stop-start		
			mid <- round(diff/2)
			cos <- start+mid
			chrs <- rep(i,length(cos))
			lib <- rep(paste("1.",lib.nums[k],sep=""),length(chrs))
			width <- stop-start		
			bind <- cbind(lib,chrs,start,stop,cos,width)	
			all <- rbind(all,bind)	
			print(dim(all))
			}
		}
	asy1.cos <- rbind(asy1.cos,all)
}
print(dim(asy1.cos))
lib.nums <- c(seq(1,81,by=1),seq(83,96,by=1))
for(k in 1:length(lib.nums)){
	print(k)
	dat <- read.table(file=paste("n96.CL.asy1hete.series2.",lib.nums[k],".bchqsnvmask.smooth.co.txt",sep=""),header=F)
	all <- NULL
	for(i in 1:5){	
		print(i)
		chr <- dat[which(dat[,2]==i),]
		if((length(chr[,1])>1)==T){
			start <- chr[,4]
			stop <- chr[,3]
			stop <- stop[-1]
			start <- start[-length(start)]
			diff <- stop-start		
			mid <- round(diff/2)
			cos <- start+mid
			chrs <- rep(i,length(cos))
			lib <- rep(paste("2.",lib.nums[k],sep=""),length(chrs))
			width <- stop-start		
			bind <- cbind(lib,chrs,start,stop,cos,width)	
			all <- rbind(all,bind)	
			print(dim(all))
			}
		}
	asy1.cos <- rbind(asy1.cos,all)
}
print(dim(asy1.cos))

mask.cos <- read.table(file="all.maskcos.txt")
hswt.cos <- mask.cos[which(mask.cos[,8]=="hs.wt"),]
wt.cos <- hswt.cos

asy1.cos <- asy1.cos

wt.tot <- 245
asy1.tot <- 191

library(GenomicRanges)
centromeres <- c(15086045,3607929,13587786,3956021,11725024)
chrs <- c("Chr1","Chr2","Chr3","Chr4","Chr5")
chr.ends <- c(30427671,19698289,23459830,18585056,26975502)
tha.cum <- cumsum(chr.ends)
tha.cum <- c(0,tha.cum)
tha.tot <- tha.cum[length(tha.cum)]
centromeres[2] <- centromeres[2]+tha.cum[2]
centromeres[3] <- centromeres[3]+tha.cum[3]
centromeres[4] <- centromeres[4]+tha.cum[4]
centromeres[5] <- centromeres[5]+tha.cum[5]

test <- seq(1,500,by=1)
j=8
ma <- rep(1,test[j])/test[j]
cos.matrix <- NULL
for(i in 1:5){
	print(i)
	wt.chr <- wt.cos[which(wt.cos[,3]==i),]
	asy1.chr <- asy1.cos[which(asy1.cos[,2]==i),]
	print(i)
	windows <- seq(1,chr.ends[i],by=300000)
	windows <- c(windows,chr.ends[i])
	cum.windows <- windows+tha.cum[i]
	print(i)
	win.wt <- NULL
	win.asy1 <- NULL
	for(j in 1:length(windows)-1){
		print(j)
		win.wt <- c(win.wt,length(which(wt.chr[,6]>=windows[j]&wt.chr[,6]<=windows[j+1])))
		win.asy1 <- c(win.asy1,length(which(as.numeric(asy1.chr[,5])>=windows[j]&as.numeric(asy1.chr[,5])<=windows[j+1])))
		}
	win.wt <- win.wt/wt.tot
	win.asy1 <- win.asy1/asy1.tot
	print(i)
	filt.wt <- filter(win.wt,ma)
	which.na <- which(is.na(filt.wt)==TRUE)
	left.na <- which.na[which(which.na<20)]
	left.val <- filt.wt[left.na[length(left.na)]+1]
	filt.wt[left.na] <- left.val
	right.na <- which.na[which(which.na>55)]
	right.val <- filt.wt[right.na[1]-1]
	filt.wt[right.na] <- right.val
	print(i)
	filt.asy1 <- filter(win.asy1,ma)
	which.na <- which(is.na(filt.asy1)==TRUE)
	left.na <- which.na[which(which.na<20)]
	left.val <- filt.asy1[left.na[length(left.na)]+1]
	filt.asy1[left.na] <- left.val
	right.na <- which.na[which(which.na>55)]
	right.val <- filt.asy1[right.na[1]-1]
	filt.asy1[right.na] <- right.val
	print(i)
	chr.dat <- cbind(rep(i,length(windows)),cum.windows,filt.wt,filt.asy1)
	cos.matrix <- rbind(cos.matrix,chr.dat)
	print(i)
}

plot(cos.matrix[,2],cos.matrix[,3],col=2,ylim=c(0,0.05),type="l",main="wt vs asy1/+ crossovers")
par(new=T)
plot(cos.matrix[,2],cos.matrix[,4],col=4,ylim=c(0,0.05),type="l")
abline(v=tha.cum,lwd=0.5)
abline(v=centromeres,lty=2,lwd=0.5)

####################
# TEL-CEN analysis #
####################

mask.cos <- read.table(file="all.maskcos.txt")
hswt.cos <- mask.cos[which(mask.cos[,8]=="hs.wt"),]
wt.cos <- hswt.cos
asy1.cos <- asy1.cos
wt.tot <- 245
asy1.tot <- 191

wins <- seq(0,1,by=0.01)
chr.ends <- c(30427671,19698289,23459830,18585056,26975502)
centromeres <- c(15086045,3607929,13587786,3956021,11725024)
chrs <- c("Chr1","Chr2","Chr3","Chr4","Chr5")

wt.left <- NULL
wt.right <- NULL
asy1.left <- NULL
asy1.right <- NULL
for(i in 1:5){
	print(i)
	wt.coords <- wt.cos[which(wt.cos[,3]==i),6]	
	left.wt <- wt.coords[which(wt.coords<centromeres[i])]
	left.prop <- left.wt/centromeres[i]
	right.arm <- chr.ends[i]-centromeres[i]	
	right.wt <- wt.coords[which(wt.coords>centromeres[i])]
	right.wt <- right.wt-centromeres[i]
	right.prop <- right.wt/right.arm
	right.prop <- 1-right.prop
	right.prop <- rev(right.prop)
	chr.left <- cbind(rep(i,length(left.prop)),left.prop)
	chr.right <- cbind(rep(i,length(right.prop)),right.prop)
	wt.left <- rbind(wt.left,chr.left)
	wt.right <- rbind(wt.right,chr.right)
	print(i)
	asy1.coords <- as.numeric(asy1.cos[which(as.numeric(asy1.cos[,2])==i),5])
	left.asy1 <- asy1.coords[which(asy1.coords<centromeres[i])]
	left.prop <- left.asy1/centromeres[i]
	right.arm <- chr.ends[i]-centromeres[i]	
	right.asy1 <- asy1.coords[which(asy1.coords>centromeres[i])]
	right.asy1 <- right.asy1-centromeres[i]
	right.prop <- right.asy1/right.arm
	right.prop <- 1-right.prop
	right.prop <- rev(right.prop)
	chr.left <- cbind(rep(i,length(left.prop)),left.prop)
	chr.right <- cbind(rep(i,length(right.prop)),right.prop)
	asy1.left <- rbind(asy1.left,chr.left)
	asy1.right <- rbind(asy1.right,chr.right)
}
wt.all <- c(wt.left[,2],wt.right[,2])
asy1.all <- c(asy1.left[,2],asy1.right[,2])
wt.wins <- NULL
asy1.wins <- NULL
for(j in 1:length(wins)-1){
	win.wt <- length(which(wt.all>=wins[j]&wt.all<wins[j+1]))
	wt.wins <- c(wt.wins,win.wt)
	win.asy1 <- length(which(asy1.all>=wins[j]&asy1.all<wins[j+1]))
	asy1.wins <- c(asy1.wins,win.asy1)
}
wt.norm <- wt.wins/wt.tot
asy1.norm <- asy1.wins/asy1.tot

####################
# TEL-CEN plotting #
####################

par(mfcol=c(1,1))
par(mar=c(1.8,1.8,1.8,1.8))
df <- 40
ylim=c(0,0.18)
plot(wins,wt.norm,type="n",ylim=ylim)
lines(smooth.spline(wins,wt.norm,df=df),col=2)
abline(h=mean(wt.norm),col=2,lty=2,lwd=0.5)
lines(smooth.spline(wins,asy1.norm,df=df),col=4)
abline(h=mean(asy1.norm),col=4,lty=2,lwd=0.5)

















############################################################################################
# intercrossover distances - analyse the same number of random positions and compare ratio #
############################################################################################

cu.wt <- read.table(file="underwood.wt.coller.txt")
hs.wt <- read.table(file="serra.wt.coller.txt")
all.wt <- rbind(cu.wt,hs.wt)
hs.sa <- read.table(file="serra.SA.coller.txt")
hs.ja <- read.table(file="serra.JA.coller.txt")
cu.hei10 <- read.table(file="underwood.HEI10.coller.txt")
cu.cmt3 <- read.table(file="underwood.cmt3.coller.txt")
hs.recq <- read.table(file="serra.recq4a4b.coller.txt")
hs.recqhei10 <- read.table(file="serra.recq4a4bHEI10.coller.txt")
cl.recqcmt3 <- read.table(file="lambing.recq4a4bcmt3.coller.txt")
cl.hei10cmt3 <- read.table(file="lambing.HEI10cmt3.coller.txt")

par(mfcol=c(2,5))
par(mar=c(1.8,1.8,1.8,1.8))

chr.ends <- c(30427671,19698289,23459830,18585056,26975502)

dat <- cu.wt
lane.cos <- NULL
lane.ran <- NULL
for(j in 1:length(unique(dat[,1]))){
	lane <- dat[which(dat[,1]==j),]
	lib.cos <- NULL
	lib.ran <- NULL
	for(k in 1:length(unique(dat[,2]))){
		lib <- lane[which(lane[,2]==dat[k,2]),]
		chr.cos <- NULL
		chr.ran <- NULL
		for(i in 1:5){
			chr <- lib[which(lib[,3]==i),]
			if(length(chr[,1]>1)){
				ran <- round(runif(length(chr[,1]),min=1,max=chr.ends[i]))
				ran <- ran[order(ran)]
				chr.ran <- c(chr.ran,diff(ran))
				chr.cos <- c(chr.cos,diff(chr[,6]))
				print("TRUE")	
				}
			}
		lib.cos <- c(lib.cos,chr.cos)
		lib.ran <- c(lib.ran,chr.ran)
		}
	lane.cos <- c(lane.cos,lib.cos)
	lane.ran <- c(lane.ran,lib.ran)
}
cu.lane.cos <- lane.cos
cu.lane.ran <- lane.ran

dat <- hs.wt
lane.cos <- NULL
lane.ran <- NULL
for(j in 1:length(unique(dat[,1]))){
	lane <- dat[which(dat[,1]==j),]
	lib.cos <- NULL
	lib.ran <- NULL
	for(k in 1:length(unique(dat[,2]))){
		lib <- lane[which(lane[,2]==dat[k,2]),]
		chr.cos <- NULL
		chr.ran <- NULL
		for(i in 1:5){
			chr <- lib[which(lib[,3]==i),]
			if(length(chr[,1]>1)){
				ran <- round(runif(length(chr[,1]),min=1,max=chr.ends[i]))
				ran <- ran[order(ran)]
				chr.ran <- c(chr.ran,diff(ran))
				chr.cos <- c(chr.cos,diff(chr[,6]))
				print("TRUE")	
				}
			}
		lib.cos <- c(lib.cos,chr.cos)
		lib.ran <- c(lib.ran,chr.ran)
		}
	lane.cos <- c(lane.cos,lib.cos)
	lane.ran <- c(lane.ran,lib.ran)
}
lane.cos <- c(lane.cos,cu.lane.cos)
lane.ran <- c(lane.ran,cu.lane.ran)
plot(density(lane.cos),col=2,ylim=c(0,max(density(lane.ran)[[2]])),xlim=c(0,30000000),main="all.wt")
abline(v=mean(lane.cos),col=2,lty=2)
par(new=T)
plot(density(lane.ran),col=4,ylim=c(0,max(density(lane.ran)[[2]])),xlim=c(0,30000000),main="")
abline(v=mean(lane.ran),col=4,lty=2)
mean(lane.cos)/mean(lane.ran)
1.240574
code <- c(rep(1,length(lane.cos)),c(rep(2,length(lane.ran))))
dco <- c(lane.cos,lane.ran)
dat <- cbind(code,dco)
wilcox <- wilcox.test(dco ~ code, data=dat)	
wilcox$p.value
1.040205e-17
mean(lane.cos)
8573616
mean(lane.ran)
6911005

dat <- hs.wt
lane.cos <- NULL
lane.ran <- NULL
for(j in 1:length(unique(dat[,1]))){
	lane <- dat[which(dat[,1]==j),]
	lib.cos <- NULL
	lib.ran <- NULL
	for(k in 1:length(unique(dat[,2]))){
		lib <- lane[which(lane[,2]==dat[k,2]),]
		chr.cos <- NULL
		chr.ran <- NULL
		for(i in 1:5){
			chr <- lib[which(lib[,3]==i),]
			if(length(chr[,1]>1)){
				ran <- round(runif(length(chr[,1]),min=1,max=chr.ends[i]))
				ran <- ran[order(ran)]
				chr.ran <- c(chr.ran,diff(ran))
				chr.cos <- c(chr.cos,diff(chr[,6]))
				print("TRUE")	
				}
			}
		lib.cos <- c(lib.cos,chr.cos)
		lib.ran <- c(lib.ran,chr.ran)
		}
	lane.cos <- c(lane.cos,lib.cos)
	lane.ran <- c(lane.ran,lib.ran)
}
lane.cos <- c(lane.cos,cu.lane.cos)
lane.ran <- c(lane.ran,cu.lane.ran)
plot(density(lane.cos),col=2,ylim=c(0,max(density(lane.ran)[[2]])),xlim=c(0,30000000),main="hs.wt")
abline(v=mean(lane.cos),col=2,lty=2)
par(new=T)
plot(density(lane.ran),col=4,ylim=c(0,max(density(lane.ran)[[2]])),xlim=c(0,30000000),main="")
abline(v=mean(lane.ran),col=4,lty=2)
mean(lane.cos)/mean(lane.ran)
1.227428
code <- c(rep(1,length(lane.cos)),c(rep(2,length(lane.ran))))
dco <- c(lane.cos,lane.ran)
dat <- cbind(code,dco)
wilcox <- wilcox.test(dco ~ code, data=dat)	
wilcox$p.value
1.873784e-16
mean(lane.cos)
8573616
mean(lane.ran)
6985027

dat <- hs.sa
lane.cos <- NULL
lane.ran <- NULL
for(j in 1:length(unique(dat[,1]))){
	lane <- dat[which(dat[,1]==j),]
	lib.cos <- NULL
	lib.ran <- NULL
	for(k in 1:length(unique(dat[,2]))){
		lib <- lane[which(lane[,2]==dat[k,2]),]
		chr.cos <- NULL
		chr.ran <- NULL
		for(i in 1:5){
			chr <- lib[which(lib[,3]==i),]
			if(length(chr[,1]>1)){
				ran <- round(runif(length(chr[,1]),min=1,max=chr.ends[i]))
				ran <- ran[order(ran)]
				chr.ran <- c(chr.ran,diff(ran))
				chr.cos <- c(chr.cos,diff(chr[,6]))
				print("TRUE")	
				}
			}
		lib.cos <- c(lib.cos,chr.cos)
		lib.ran <- c(lib.ran,chr.ran)
		}
	lane.cos <- c(lane.cos,lib.cos)
	lane.ran <- c(lane.ran,lib.ran)
}
plot(density(lane.cos),col=2,ylim=c(0,max(density(lane.ran)[[2]])),xlim=c(0,30000000),main="hs.sa")
abline(v=mean(lane.cos),col=2,lty=2)
par(new=T)
plot(density(lane.ran),col=4,ylim=c(0,max(density(lane.ran)[[2]])),xlim=c(0,30000000),main="")
abline(v=mean(lane.ran),col=4,lty=2)
mean(lane.cos)/mean(lane.ran)
1.21
code <- c(rep(1,length(lane.cos)),c(rep(2,length(lane.ran))))
dco <- c(lane.cos,lane.ran)
dat <- cbind(code,dco)
wilcox <- wilcox.test(dco ~ code, data=dat)	
wilcox$p.value
3.699823e-07
mean(lane.cos)
8455422
mean(lane.ran)
6946373


############################################################################
# seperate into pericentromeres/centromeres versus arms - count crossovers #
############################################################################

cu.wt <- read.table(file="underwood.wt.coller.txt")
hs.wt <- read.table(file="serra.wt.coller.txt")
all.wt <- rbind(cu.wt,hs.wt)
hs.sa <- read.table(file="serra.SA.coller.txt")
hs.ja <- read.table(file="serra.JA.coller.txt")
cu.hei10 <- read.table(file="underwood.HEI10.coller.txt")
cu.cmt3 <- read.table(file="underwood.cmt3.coller.txt")
hs.recq <- read.table(file="serra.recq4a4b.coller.txt")
hs.recqhei10 <- read.table(file="serra.recq4a4bHEI10.coller.txt")
cl.recqcmt3 <- read.table(file="lambing.recq4a4bcmt3.coller.txt")
cl.hei10cmt3 <- read.table(file="lambing.HEI10cmt3.coller.txt")
cu.wt.n <- 192
hs.wt.n <- 245
all.wt.n <- cu.wt.n+hs.wt.n
hs.sa.n <- 192
hs.ja.n <- 172
cu.hei10.n <- 192
cu.cmt3.n <- 384
hs.recq.n <- 191
hs.recqhei10.n <- 192
cl.recqcmt3.n <- 192
cl.hei10cmt3.n <- 192

chr.ends <- c(30427671,19698289,23459830,18585056,26975502)
chrs <- c("Chr1","Chr2","Chr3","Chr4","Chr5")
cen.coords <- read.csv(file="cen.pericen.coords.csv")

dat <- all.wt
dat.ann <- NULL
for(i in 1:5){
	print(i)
	chr.cen <- cen.coords[i,]
	chr.dat <- dat[which(dat[,3]==i),6]
	chr.dat <- chr.dat[order(chr.dat)]
	chr.dat <- cbind(chr.dat,rep(0,length(chr.dat)),rep(i,length(chr.dat)))
	dat.coords <- as.numeric(chr.dat[,1])
	chr.dat[which(dat.coords>=1&dat.coords<=chr.cen[1,3]),2] <- 1
	chr.dat[which(dat.coords>=chr.cen[1,3]&dat.coords<=chr.cen[1,4]),2] <- 2
	chr.dat[which(dat.coords>=chr.cen[1,4]&dat.coords<=chr.cen[1,5]),2] <- 3
	chr.dat[which(dat.coords>=chr.cen[1,5]&dat.coords<=chr.cen[1,6]),2] <- 2
	chr.dat[which(dat.coords>=chr.cen[1,6]&dat.coords<=chr.ends[i]),2] <- 1
	dat.ann <- rbind(dat.ann,chr.dat)
	print(dim(dat.ann))
}
dat.count <- NULL
for(i in 1:5){
	chr.dat <- dat.ann[which(dat.ann[,3]==i),]
	count <- c(length(chr.dat[which(chr.dat[,2]==1),2]),length(chr.dat[which(chr.dat[,2]==2),2]),length(chr.dat[which(chr.dat[,2]==3),2]))
	dat.count <- cbind(dat.count,count)
}
dat.count <- cbind(dat.count,rowSums(dat.count))
dat.count <- rbind(dat.count,colSums(dat.count))
write.csv(dat.count,file="summary.allwt.csv")

dat <- hs.wt
dat.ann <- NULL
for(i in 1:5){
	print(i)
	chr.cen <- cen.coords[i,]
	chr.dat <- dat[which(dat[,3]==i),6]
	chr.dat <- chr.dat[order(chr.dat)]
	chr.dat <- cbind(chr.dat,rep(0,length(chr.dat)),rep(i,length(chr.dat)))
	dat.coords <- as.numeric(chr.dat[,1])
	chr.dat[which(dat.coords>=1&dat.coords<=chr.cen[1,3]),2] <- 1
	chr.dat[which(dat.coords>=chr.cen[1,3]&dat.coords<=chr.cen[1,4]),2] <- 2
	chr.dat[which(dat.coords>=chr.cen[1,4]&dat.coords<=chr.cen[1,5]),2] <- 3
	chr.dat[which(dat.coords>=chr.cen[1,5]&dat.coords<=chr.cen[1,6]),2] <- 2
	chr.dat[which(dat.coords>=chr.cen[1,6]&dat.coords<=chr.ends[i]),2] <- 1
	dat.ann <- rbind(dat.ann,chr.dat)
	print(dim(dat.ann))
}
dat.count <- NULL
for(i in 1:5){
	chr.dat <- dat.ann[which(dat.ann[,3]==i),]
	count <- c(length(chr.dat[which(chr.dat[,2]==1),2]),length(chr.dat[which(chr.dat[,2]==2),2]),length(chr.dat[which(chr.dat[,2]==3),2]))
	dat.count <- cbind(dat.count,count)
}
dat.count <- cbind(dat.count,rowSums(dat.count))
dat.count <- rbind(dat.count,colSums(dat.count))
write.csv(dat.count,file="summary.hswt.csv")


